Fix --failed-tests, Unittests

milljm opened the pull request:
Create unittest for --failed-tests. While doing so,
discovered a rather nasty bug:

Fix the potential for unittests to fail, and not reporting
it. We were returning error codes, but not setting the
status of the test

Allow TestHarness to use a different naming convention for
.previous_test_results.json. This was necessary to prevent
the many-failing-tests in unittests from clobbering the
results file during the crucial --failed-tests test.

Jeez... fix all the dumb naming conventions of results_file,
results_storage... was overly confusing with similarly named
variables. results_file is a file object. results_storage is
json object.

Closes #14512 #14575

Commit:
Fix --failed-tests, Unittests

Create unittest for --failed-tests. While doing so,
discovered a rather nasty bug:

Fix the potential for unittests to fail, and not reporting
it. We were returning error codes, but not setting the
status of the test

Allow TestHarness to use a different naming convention for
.previous_test_results.json. This was necessary to prevent
the many-failing-tests in unittests from clobbering the
results file during the crucial --failed-tests test.

Jeez... fix all the dumb naming conventions of results_file,
results_storage... was overly confusing with similarly named
variables. results_file is a file object. results_storage is
json object.

Closes #14512 #14575

Commit:
Fix the odd logic of unittest

While it was passing, I think I messed up the assert logic.

milljm commented on python/TestHarness/testers/RunApp.py:
```diff
@@ -254,16 +254,16 @@ def testExitCodes(self, moose_dir, options, output):
             elif self.exit_code != 0 and specs['should_crash'] == False:
                 # Let's look at the error code to see if we can perhaps further split this out later with a post exam
                 reason = 'CRASH'
-                return "\n\nExit Code: " + str(self.exit_code)
```
This is what was causing TestHarness unittests to pass. We were exiting this function before establishing a status for the test.

permcody commented on python/TestHarness/testers/RunApp.py:
```diff
@@ -254,16 +254,16 @@ def testExitCodes(self, moose_dir, options, output):
             if reason != '':
-                self.setStatus(self.fail, reason)
+                self.setStatus(self.fail, str(reason))
+                reason += "\n\nExit Code: " + str(self.exit_code)
```
Hmm - are you sure you want to overload the meaning of "reason" here? I can see that this logic is OK because you are setting the status then changing the value of reason. This is not a very good design because if something gets reordered (think about what would happen if the setStatus method got moved around), the reason variable might contain the "reason" AND some extra newlines and an exit code. How about creating a second local variable that you use to hang onto any extra stuff that you want to print out? Or you could just return that stuff directly.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/451008/) on 5a4bf9c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14583/site)

This comment will be updated on new commits.

permcody commented on python/TestHarness/testers/RunApp.py:
```diff
@@ -280,6 +276,11 @@ def processResults(self, moose_dir, options, output):
+        reason = self.testExitCodes(moose_dir, options, output)
+        if reason != '':
+            self.setStatus(self.fail, reason)
+            output += "\n\nExit Code: " + str(self.exit_code)
```
I thought we weren't going to try to refactor this just yet. This isn't going to work like you want because processResults() is a virtual method. It's overridden in most of the Tester objects. Many of the Tester objects are assuming that TestExitCodes() is going to return extra stuff to append to the output (the existing behavior) but now you are returning reason. 

If you are going to refactor, you have to change every Tester object, or you can just do that in a separate PR.

Commit:
Nope. Try again

For now, due to critical label, this will solve the issue for now.
Leaving us time to perform a proper refactor.

Refs #14512
